### PR TITLE
Help chat

### LINKED ISSUE #6610: Enso Chat
## General idea

The Enso Chat is a built-in the app chat for the clients. In case a client does not know how to perform a task, they can click it and instantly chat with someone from our team.

<br/>

## Accessing the chat

You can access it by pressing the chat icon at the top-right corner:

image

After the icon is pressed, a chat window should open. It should be placed on the right side of the component browser and the dashboard, so the user would be able to modify the graph and use the dashboard while keeping the chat opened. However, as WebGL scene resizing is slow, technically, it should float over the EnsoGL scene and we should expose an API from GUI allowing for "virtual clipping of the scene". Such a virtual clipping would allow for passing information to EnsoGL about the visible scene size, which would, for example, prevent showing component browser off-screen (under the chat). However, such an API should not be part of the first implementation task.

image

<br/>

## Components of the chat window

image

The chat window should display the following properties (top to bottom):
- The name of the current thread. In case no thread was created earlier, the name should be "New chat thread". After clicking the thread name, list of threads is displayed. In case in any of the threads there are new messages, an indicator (red bubble) should be shown next to the thread drop-down.
- Conversation divided into chat bubbles. Each chat bubble contains avatar of the person who has written the message, time formatted in the user-local format, and the message. For the Enso team, the avatar should be taken from Enso Discord server, and the real, full name should be used (first and last name). This makes the client feel that they are really speaking with a real person.
- Under the last message of Enso Team member, a reaction panel should be displayed, allowing the user to react with an emoji with a single mouse click. This also encourages the user to react to messages, as it is so easily accessible.
- After moving the mouse over any of Enso Team member messages, a reaction panel hovering over the bottom part of the message should be displayed as well. If there is at least one reaction given, the reaction panel should stay visible in the chat history under the message.
- An input panel allowing typing new messages. If the user typed at least one letter, the buttons under the input field start to be active (less transparent). The "reply" button simply posts the message, while the "New question? Click to start a new thread!" button posts this question in a new thread and switches to that thread.
- In case the user is not a paid user, a notification should be shown, as presented on the above image.

<br/>

## How the chat should work under the hood?
Under the hood, the chat should speak with a Enso Customer Chat Discord Bot (which should be created as part of this task), which should be a bridge between this chat and our Discord server. An example user session should look like follows:
- As soon as the user logs in to Enso, the information about user login should be send to Enso Customer Chat Discord Bot and the chat history is fetched. In case new messages were posted since user checked them last time, an indicator (red bubble) should be displayed next to the chat icon.
- After the user clicks the chat icon, the chat window opens with the last selected thread. In case the user did not use the chat yet, a new thread is created with the default name. 
- As soon as a thread in the Enso Chat is created, a thread in the "Customer Chat" Discord channel (of type "Forum") is created. The title of the Discord thread should be `name of the channel [customer_id]`.
- Every message posted by the user should be replicated and posted by the bot in the appropriate Discord thread. As soon as any of Enso team members replies in the thread, the message should be sent by bot to the application, if it is opened. The application should not waste any CPU cycles waiting for the message, so instead of polling, we should use web-sockets with direct communication between the chat bot and the Enso app. 
- The communication with Discord should be done via [discord.js library](https://discord.js.org), as it is the most popular of all Discord bindings.
- The emojis should use the open-source Twitter emoji images: https://twemoji.twitter.com
- The backend should run on Heroku-like thing. Simple git-push should be able to update the bot and its development/deployement should be as straightforward as possible. Please compare available options and chose one that is popular, robust, and provides fast, git-based development speed.

<br/>

## Possible future extensions

This chat and bot can (and probably will) become the basis for our customer support. For example, if a logged customer (via app our our website) asks a question about their billing, someone from our team should be able to write in the Discord thread `!billing history` and get a billing history printed in this discord thread (visible only for Enso members). This way, our customer support could really fast query the bot for information about specific user and reply to them from within the same chat window. This could also allow for creating template replies. For example, someone from our customer support team could write `!send pricing info` and a nicely formatted message about pricing info will be send to the user.

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
Closes #6610

### Important Notes
- Should be tested alongside https://github.com/enso-org/enso-bot/pull/1.
- Unread messages notification has not been implemented

### Screencasts
[screen-recorder-mon-jul-03-2023-20-29-41.webm](https://github.com/enso-org/enso/assets/4046547/f1a34aba-8add-4bbf-a7ac-81ad0d06f471)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/chat.tsx [hunk 1]
```diff
@@ -0,0 +1,941 @@
...
+ * This SHOULD be the same limit as the chat backend (the maximum number of messages sent in
+ * `serverThread` events). */
+const MAX_MESSAGE_HISTORY = 25
+
+// =============
+// === Types ===
+// =============
+
+/** All possible reaction emojis. */
+type ReactionSymbol = '❤️' | '🎉' | '👀' | '👍' | '👎' | '😀' | '🙁'
```
Comment (wdanilo) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/chat.tsx]: this is a repeated code - above we have the same. Can we refactor it, please?

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/chat.tsx [hunk 2]
```diff
@@ -0,0 +1,941 @@
...
+
+// =====================
+// === Message Types ===
+// =====================
+
+// FIXME[sb]: Consider deduplicating.
```
Comment (wdanilo) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/chat.tsx]: This should be part of cleaning this PR if its about code deduplication :) if its about anything else, it needs better docs.

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/chat.tsx [hunk 3]
```diff
@@ -0,0 +1,941 @@
...
+
+/** Props for a {@link Chat}. */
+export interface ChatProps {
+    /** This should only be false when the panel is closing. */
+    isOpen: boolean
+    doClose: () => void
+}
+
+/** Chat sidebar. */
+function Chat(props: ChatProps) {
```
Comment (wdanilo) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/chat.tsx:375]: this function is 400 lines long! Can we refactor it to several well named functions instead, please?

**Review by wdanilo — APPROVED** [review]
This is super cool! A few questions:
1. Where is server implementation? We should keep it in our repo as well.
2. The design is not exactly as in Figma, do we plan to fix it in this PR or in a separate one?
3. Where on Discord the user messages land? Do we have access to it, or it is in a test server for now?

**Comment by somebody1234** [comment]
- server implementation is linked in the original post. can be easily moved into this repo, but thinking about it it might make deploy on commit a bit harder unless the app hosting platform we use supports sparse checkouts... or actually we could use CI to build to a separate repo. thoughts?
- design not identical to figma - whoops, forgot to bring this up. I don't think i have a link to the figma, and the screenshot seems like it's scaled, likely because of retina displays, so its a bit difficult for me to get the right spacings (etc)
- the discord messages land in a test server, yes - see the linked PR about the server impl for details
- deduplication: currently the duplicated code is across two different repos, and that's the part making it hard to deduplicate. if you have any ideas that'd be great
-  repeated code: fair point, originally i refactored it *to* this to make it consistent with the backend, but yeah the constant can be copied to the backend instead. it won't actually be used in the backend but itll be consistent, *and* there will be no repetition

**Comment by somebody1234** [comment]
400 line long function: I'll see whst i can do

**Comment by wdanilo** [comment]
- @somebody1234 oh, sorry. Keeping it in a separate repo makes sense. I wanted to tell "this organization", not necessarily this repo, sorry for confusion. I totally missed that link!
- Link to Figma: https://www.figma.com/file/YdrxRRNoZWqePEJQye80YG/ide2-v49.8?type=design&node-id=0%3A1&mode=design&t=1qh6VVj0wKvvfddF-1
- Regarding duplicated code across repos – hmmmm. What about making that repo an NPM library? Just an idea, probably not best, I'm asking about your opinion here to chat about possibilities.
- Regarding long functions - I'd be thankful if you could try splitting them down, it would make the code way more maintainable long term :)

**Comment by somebody1234** [comment]
re: using a npm library - it would probably work, but upon some investigation it seems like it's possible to use [git dependencies](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#git-urls-as-dependencies), which is probably more convenient, since it wouldn't require publishing new tags for local development

also, it's not like it should be publicly visible on npm - not that it *can't* be, but i do think what determines whether something belongs on npm should be an intentional decision, rather than "we have to put it on npm for things to work" - when there are other options, at least

**Comment by wdanilo** [comment]
Oh, I did not know that nom supports git dependencies. This is way better solution, let's use it! :)

**Comment by PabloBuchu** [comment]
Please adjust spaces around inputs to match figma

Screenshot 2023-07-06 at 12 23 26

**Comment by PabloBuchu** [comment]
@somebody1234 
1. The arrow in threads list is not animated. Should be toggled (down for rolled up; up for expanded)
2. The emoji menu should be visible only on hover.
3. Emoji `...` at the end with full emoji menu? This one please estimate if its doable at all and how hard would it be.
4. Creating new thread should append +x to the default name `New chat thread 1`. Right now renaming threads doesn't work on mac and I can follow only latest thread. Previously created are gone. 
4.a Weirdly threads appeared after restarting enso but creating new ones do not appended them to thread list
5. No highlighting when moving mouse over the thread list. The current opened should be highlighted but also the one I have my pointer over
6. Thread renaming. When you hover your mouse over the thread name the mouse pointer should be changed to `text` `ⵊ` and after click name should be changed to input widget with save on enter
7. Also on thread list we should have 1 additional icon on the left (if its doable):
- trash for delete
8. With expanded thread list clicking outside it does not roll it up
9.  padding left / right should be the same as top and down for "New question?" and "Reply"
Screenshot 2023-07-07 at 11 27 14
Screenshot 2023-07-07 at 11 27 55

10. Bug with opened workspace:
Screenshot 2023-07-07 at 11 35 13

11. I also have double cursor pointers when working as above (can't show on screenshot) 
12. There is no help chat on workspace view (needed to go back to dashboard and leave opened to have it on project view). I guess this one is not doable without either having it in rust or having header in react right?

**Comment by somebody1234** [comment]
1. done locally
2. not sure how this should work
3. sounds reasonable, naively it should just be a scrolling div with all the emoji - so like the current reaction bar, just with more emojis. one problem though, would be whether it's needed - i'd imagine some (or many) emojis should be omitted, either because they aren't professional (e.g. the vomiting emoji), or because they are rarely useful (e.g. the regional indicator emojis (🇦-🇿))
4. ~~will do~~ done locally
5. done locally
6. done locally
7. tbd
8. done locally
9. done locally (visually). they actually were the same size - it's because of the rounded corners that it looks like they're different sizes
10. done locally
11. cannot repro, i'm assuming fix for 10 has fixed this too
12. yeah. i think the simplest way is to do things the same way as the "back to ide" event - have rust implement the ui to open it, and then use events to send information back and forth

**Comment by PabloBuchu** [comment]
still having issues that threads are not appending to the list but only current is visible. now even reopening doesn't help

https://github.com/enso-org/enso/assets/2855109/0dc53011-32de-40e4-9f44-8255c6d0feaf

**Comment by PabloBuchu** [comment]
@somebody1234 
1. lets not add more emojis for now. can we just fix emoji menu on hover? currently you can only like the recent response
2. can we add a red dot on help chat blue button whenever there is unread message? how hard would it be?

**Comment by somebody1234** [comment]
1. will do tomorrow
2. on it. seems like it's harder than i expected... i'm thinking of using http + polling, because i'd imagine most users would rarely (if ever) use the help chat

**Comment by somebody1234** [comment]
i just realized, i don't think this works with multiple users in an organization. i'll have to take a look at that sometime

**Comment by PabloBuchu** [comment]
> i just realized, i don't think this works with multiple users in an organization. i'll have to take a look at that sometime

this one is not obvious. lets fix conflicts and open for review and merge (deploy) this implementation then we see

**Review by PabloBuchu — APPROVED** [review]

**Pull request merged**